service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java

```diff
@@ -55,68 +52,91 @@ public PaymentHandler(String payoneInterfaceName, String tenantName,
-
+            return attemptToProcessPayment(paymentId, 0);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            return handleConcurrentModificationException(paymentId, concurrentModificationException);
         } catch (final NotFoundException | NoCartLikeFoundException e) {
             return handleNotFoundException(paymentId, e);
         } catch (final ErrorResponseException e) {
             return errorResponseHandler(e, paymentId);
-        } catch (final CompletionException e) {
-            return completionExceptionHandler(e, paymentId);
         } catch (final Exception e) {
-            return handleThrowableInResponse(e, paymentId);
+            return handleException(e, paymentId);
+        }
+    }
+
+    private PaymentHandleResult attemptToProcessPayment(
+        @Nonnull final String paymentId,
+        final int currentAttemptCount) throws InterruptedException {
+
+        try {
+            return processPayment(paymentId);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            if (currentAttemptCount == RETRIES_LIMIT - 1) {
+                throw concurrentModificationException;
+            } else {
+                Thread.sleep(RETRY_DELAY);
+                return attemptToProcessPayment(paymentId, currentAttemptCount + 1);
+            }
         }
     }
 
-     private PaymentHandleResult handleNotFoundException(@Nonnull String paymentId, @Nonnull Exception exception ) {
+    private PaymentHandleResult processPayment(@Nonnull final String paymentId) {
+        final PaymentWithCartLike paymentWithCartLike =
+            commercetoolsQueryExecutor
+                .getPaymentWithCartLike(paymentId);
+
+        final String paymentInterface = paymentWithCartLike
+            .getPayment()
+            .getPaymentMethodInfo()
+            .getPaymentInterface();
+
+        if (!payoneInterfaceName.equals(paymentInterface)) {
+            logger.warn(tenantNameKeyValue,
+                "Wrong payment interface name: expected [{}], found [{}]",
+                payoneInterfaceName, paymentInterface);
+            return new PaymentHandleResult(HttpStatusCode.BAD_REQUEST_400);
+        }
+
+        paymentDispatcher.dispatchPayment(paymentWithCartLike);
+        return new PaymentHandleResult(HttpStatusCode.OK_200);
+    }
+
+    private PaymentHandleResult handleConcurrentModificationException(
+        @Nonnull final String paymentId,
+        @Nonnull final ConcurrentModificationException concurrentModificationException) {
+
+        final String errorMessage = format("The payment with id '%s' couldn't be processed after %s retries.",
+            paymentId, RETRIES_LIMIT);
+        logger.error(errorMessage, concurrentModificationException);
+        return new PaymentHandleResult(HttpStatusCode.ACCEPTED_202, errorMessage);
```

Comment:
How does it work in this case? 202 HTTP status is returned to user, so it means at some point later it will be processed? Who will process it later?

Reply:
Actually I am also not sure here because it was already like that. @butenkor any idea why it used to return 202 here? When will it be acted upon later?

> The HyperText Transfer Protocol (HTTP) 202 Accepted response status code indicates that the request has been received but not yet acted upon.

Reply:
There are 2 build-in cron jobs which process not processed payments. Those run independently from handle request called externally. It means that after the caller got a 202 a minute later a cron could successfully process the payment.

Reply:
Do you think it makes sense to add this info to the log message?